internal/app/init/pkg/system/events/events.go

```diff
@@ -0,0 +1,76 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package events
+
+import "time"
+
+// MaxEventsToKeep is maximum number of events to keep per service before dropping old entries
+const MaxEventsToKeep = 64
+
+// ServiceState is enum of service run states
+type ServiceState int
+
+// ServiceState constants
+const (
+	StateInitialized ServiceState = iota
+	StatePreparing
+	StateWaiting
+	StateRunning
+	StateStopping
+	StateFinished
+	StateFailed
+)
+
+func (state ServiceState) String() string {
+	switch state {
+	case StateInitialized:
+		return "Initialized"
+	case StatePreparing:
+		return "Preparing"
+	case StateWaiting:
+		return "Waiting"
+	case StateRunning:
+		return "Running"
+	case StateStopping:
+		return "Stopping"
+	case StateFinished:
+		return "Finished"
+	case StateFailed:
+		return "Failed"
+	default:
+		return "Unknown"
+	}
+}
+
+// ServiceEvent describes state change of the running service
+type ServiceEvent struct {
+	Message   string
+	State     ServiceState
+	Timestamp time.Time
+}
+
+// ServiceEvents is a fixed length history of events
+type ServiceEvents struct {
+	Events    []ServiceEvent
+	Discarded int
+}
+
+// Push appends new event to the history popping out oldest event on overflow
+func (events *ServiceEvents) Push(event ServiceEvent) {
+	events.Events = append(events.Events, event)
+	if len(events.Events) > MaxEventsToKeep {
+		// TODO: this might be leaking memory on the elements pushed out
+		//       this should be replaced with real circular buffer
```

Comment:
You made me curious -- https://yourbasic.org/golang/implement-fifo-queue/ mentions 
```
// Dequeue
queue[0] = "" // Erase element (write zero value)
queue = queue[1:]
```
or switching to https://golang.org/pkg/container/list/. 

Should we do `events.Events[0] = nil` or `events.Events[0] = ServiceEvent{}`  ?

Reply:
not Go expert here, and I didn't do any experiments. so just my 2 cents below:

slice should be a pointer to underlying array. when we grow slice by appending more items, array is also growing (in increments, not for each append), but slice always points to the "visible" part of the array. when we move slice start, I don't know if Go is able to release parts of the array which are not referenced by any slice out there.

so in the end probably when the array is re-allocated for new events, it will be copied to a new memory location, and that would clear old entries

but I didn't like that, so I've rewritten that with kind of ring buffer

Reply:
> Should we do events.Events[0] = nil or events.Events[0] = ServiceEvent{}

we can't do `nil`, as it's slice of structs (not pointers to structs). no pointers == less GC, so adding pointers imho doesn't help

assigning null struct doesn't help much either, as struct still occupies the space (and there are no pointers in the struct itself).

I wanted to be on a safe side here. In general we shouldn't have many events. but I thought what if we have a service which is constantly crashing/restarting - I would probably want to make sure `init` doesn't blow up on memory if it runs for 1 year without being restarted